File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool canship(vector<int>& weights, int days, int capacity) {

        int currentload = 0;
        int required_days = 1;

        for (int x : weights) {
            if (currentload + x > capacity) {
                required_days++;
                currentload = 0;
            }
            currentload += x;
        }
        return required_days <= days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int low = *max_element(weights.begin(), weights.end());
        int high = accumulate(weights.begin(), weights.end(),0);
         int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canship(weights, days, mid)) {
                ans = mid;
                high=mid-1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
import java.util.*;

class Solution {

    // Helper method to check if capacity is enough
    public boolean canShip(int[] weights, int days, int capacity) {
        int currentLoad = 0;
        int requiredDays = 1;

        for (int weight : weights) {
            if (currentLoad + weight > capacity) {
                requiredDays++;
                currentLoad = 0;
            }
            currentLoad += weight;
        }

        return requiredDays <= days;
    }

    public int shipWithinDays(int[] weights, int days) {
        int low = Arrays.stream(weights).max().getAsInt();     // min possible capacity
        int high = Arrays.stream(weights).sum();               // max possible capacity

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canShip(weights, days, mid)) {
                high = mid - 1; // Try a smaller capacity
            } else {
                low = mid + 1;  // Need more capacity
            }
        }

        return low;  // The minimum capacity that works
    }
}
